File: Python/ByRound/1039/1039_A_Timetable.py
# Problem: CF 1039 A - Timetable
# https://codeforces.com/contest/1039/problem/A

"""
Algorithm: Difference Array + Greedy Construction
Time Complexity: O(n)
Space Complexity: O(n)

This solution uses a difference array technique to determine whether a valid timetable exists,
and then greedily constructs a valid timetable for stop B.

The key idea is to use difference array to track the constraints on the arrival times.
For each bus that departs at time a[i], it can arrive at position x[i]. We process the constraints
and ensure that the resulting sequence b is valid and satisfies all conditions.

The algorithm works by:
1. Using a difference array S to represent constraints on valid arrival positions
2. Checking if the constraints are feasible (non-negative prefix sum)
3. Constructing the b array greedily from right to left with appropriate adjustment rules
4. Validating the final constructed timetable
"""

import sys

input = lambda: sys.stdin.readline()[:-1]
ni = lambda: int(input())
na = lambda: list(map(int, input().split()))
yes = lambda: print("yes")
Yes = lambda: print("Yes")
YES = lambda: print("YES")
no = lambda: print("no")
No = lambda: print("No")
NO = lambda: print("NO")

n, t = na()
a = na()
x = na()

# Difference array to track valid arrival position constraints
S = [0] * (1 + n)
for i in range(n):
    S[x[i]] -= 1  # Mark that at position x[i], we have one less valid arrival
    S[i + 1] += 1  # Mark that after position i+1, we gain one more valid arrival
# Compute prefix sum to get actual constraint counts
for i in range(n):
    S[i + 1] += S[i]

# Check if constraint is feasible and x is non-decreasing
if all(x >= 0 for x in S) and all(x[i] <= x[i + 1] for i in range(n - 1)):
    b = [-1] * n  # Initialize b array
    FLAG = 1  # Flag to check special conditions
    
    # Process from right to left
    for i in range(n - 1, -1, -1):
        # Case 1: Same x value as previous bus & previous bus has a smaller x
        if i > 0 and x[i - 1] == x[i] and (i == n - 1 or x[i] < x[i + 1]):
            if x[i] < n and a[x[i]] - a[x[i] - 1] == 1:
                FLAG = 0  # Special condition violated
            b[i] = a[x[i] - 1] + 1  # Set b[i] to a[x[i]-1]+1
        elif i < n - 1 and x[i] == x[i + 1]:
            # Case 2: Same x value as next bus - set b[i] to be one less than next
            b[i] = b[i + 1] - 1
        else:
            # Case 3: Normal case - set b[i] to a[x[i]-1]
            b[i] = a[x[i] - 1]
    
    # Final validation
    if FLAG and all(b[i] < b[i + 1] for i in range(n - 1)):
        Yes()
        print(*[x + t for x in b])  # Add t to get actual arrival times
    else:
        No()
else:
    No()


# https://github.com/VaHiX/CodeForces/

File: CPP/ByTier/A/1039_A_Timetable.cpp
// Problem: CF 1039 A - Timetable
// https://codeforces.com/contest/1039/problem/A

/*
Algorithm/Technique: 
This problem involves constructing a valid timetable for bus arrivals given departure times and constraints on arrival order. 
The approach uses a difference array technique to track how many buses can arrive at each position, and then assigns arrival times accordingly.
The key idea is to ensure that each bus arrives at least 't' units of time after departure and that the maximum arrival position for each bus is respected.

Time Complexity: O(n)
Space Complexity: O(n)

*/
#include <stdio.h>
#include <ctype.h>
#include <stdio.h>
#include <algorithm>
#include <algorithm>

using namespace std;
typedef long long LL;
const int RLEN = 1 << 18 | 1;
inline char nc() {
  static char ibuf[RLEN], *ib, *ob;
  (ib == ob) && (ob = (ib = ibuf) + fread(ibuf, 1, RLEN, stdin));
  return (ib == ob) ? -1 : *ib++;
}
inline LL rd() {
  char ch = nc();
  LL i = 0, f = 1;
  while (!isdigit(ch)) {
    if (ch == '-')
      f = -1;
    ch = nc();
  }
  while (isdigit(ch)) {
    i = (i << 1) + (i << 3) + ch - '0';
    ch = nc();
  }
  return i * f;
}
const int N = 2e5 + 50;
int n, c[N], d[N];
LL t, a[N], x[N];
int main() {
  n = rd(), t = rd();
  for (int i = 1; i <= n; i++)
    a[i] = rd();
  for (int i = 1; i <= n; i++)
    if ((x[i] = rd()) < x[i - 1] || x[i] < i) {
      puts("No");
      return 0;
    }
  // Difference array technique to count how many buses can arrive at each position
  for (int i = 1; i <= n; i++)
    ++c[i], --c[x[i]], d[x[i]] = 1;
  d[n] = 0;
  for (int i = 1; i <= n; i++) {
    c[i] += c[i - 1];
    if (c[i] && d[i]) {
      puts("No");
      return 0;
    }
    // Set minimum valid arrival time based on next departure
    if (c[i])
      x[i] = a[i + 1] + t;
    // Ensure strictly increasing order of arrival times
    x[i] = max(x[i - 1] + 1, x[i]);
    // Ensure time constraint (each bus arrives at least t time units after departure)
    x[i] = max(x[i], a[i] + t);
    // Check if arrival time conflicts with previous constraints
    if (d[i] && x[i] >= a[i + 1] + t) {
      puts("No");
      return 0;
    }
  }
  puts("Yes");
  for (int i = 1; i <= n; i++)
    printf("%I64d ", x[i]);
}


// https://github.com/VaHiX/CodeForces/